tests/utilities/test_collections.py

```diff
@@ -328,6 +328,19 @@ class RandomPydantic(pydantic.BaseModel):
             output_model.val == input_model.val
         ), "The fields value should be used, not the default factory"
 
+    def test_visit_collection_remembers_unset_pydantic_fields(self):
+        class RandomPydantic(pydantic.BaseModel):
+            val: uuid.UUID = pydantic.Field(default_factory=uuid.uuid4)
+
+        input_model = RandomPydantic()
+        output_model = visit_collection(
+            input_model, visit_fn=visit_even_numbers, return_data=True
+        )
+
+        assert output_model.json(exclude_unset=True) == input_model.json(
+            exclude_unset=True
+        ), "Unset fields values should be remembered and preserved"
```

Comment:
I usually like to have my tests be as "flat" and hardcoded as I can make them, so they're dead simple to reason about.  What if we do something like this (assuming that we've extended the model to have, say, three string fields called `val1`, `val2`, `val3`:

```suggestion
        input_model = RandomPydantic(val1="first", val2="second")
        assert input_model.json(exclude_unset=True) == {
            "val1": "first-value",
            "val2": 'second"
        }
        
        output_model = visit_collection(
            input_model, visit_fn=visit_even_numbers, return_data=True
        )
        assert output_model.json(exclude_unset=True) == {
            "val1": "first",
            "val2": 'second"
        }, "Unset fields values should be remembered and preserved"
```

This way we're expanding the values so the person reading the test can see the intent directly.  Like, if you get an assertion failure saying that you expected

```
{"val1": "first", "val2": "second"}
```

and got :
```
{"val1": "first", "val2": "second", "val3": null}
```

It will be hyper-clear what's going on.

Reply:
I __strongly dislike__ staring at failing test output so I'm for tests being hyper clear about failures. I took your advice from above and parameterized this test which makes it a bit clunky to assert on the `input_model`. The assertion on the `output_model` did become more clear because I was able to assert on the input dict directly. What do you think about that?

Reply:
Oooh I like that a lot!  More tests for the buck!